### PR TITLE
RavenDB-10294 Add a thread pool that will server as object pool of lo…

### PR DESCRIPTION
**Pull request by maximburyak**
…ng running dedicated threads

### CODE REVIEW

File: src/Raven.TestDriver/Raven.TestDriver.csproj [hunk 1]
```diff
@@ -1,6 +1,6 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
-    <Version></Version>
+    <Version>4.0.0-custom-40</Version>
```
Comment (ppekrol) [inline src/Raven.TestDriver/Raven.TestDriver.csproj]: can you revert this?

File: src/Raven.Client/Raven.Client.csproj [hunk 2]
```diff
@@ -1,6 +1,6 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
-    <Version></Version>
+    <Version>4.0.0-custom-40</Version>
```
Comment (ppekrol) [inline src/Raven.Client/Raven.Client.csproj]: can you revert this?

File: src/Raven.Client/Properties/VersionInfo.cs [hunk 3]
```diff
@@ -3,7 +3,7 @@
 using System.Reflection;
 using Raven.Client.Properties;
 
-[assembly: RavenVersion(Build = "40", CommitHash = "a377982", Version = "4.0", FullVersion = "4.0.0-custom-40")]
+[assembly: RavenVersion(Build = "40", CommitHash = "96b308e", Version = "4.0", FullVersion = "4.0.0-custom-40")]
```
Comment (ppekrol) [inline src/Raven.Client/Properties/VersionInfo.cs]: can you revert this?

**Comment by maximburyak** [comment]
removed files that were modified during build

File: src/Raven.Client/Documents/Conventions/JsonNetBlittableEntitySerializer.cs [hunk 4]
```diff
@@ -25,6 +25,11 @@ public JsonNetBlittableEntitySerializer(DocumentConventions conventions)
             _generateEntityIdOnTheClient = new GenerateEntityIdOnTheClient(conventions, null);
         }
 
+        public static void CleanThreadStatics()
```
Comment (ayende) [inline src/Raven.Client/Documents/Conventions/JsonNetBlittableEntitySerializer.cs:28]: I don't particularly like this approach. It means that we spread the knowledge about clearing thread local state far and wide. We need to register this as an event, instead.

File: src/Raven.Server/Documents/ETL/EtlProcess.cs [hunk 5]
```diff
@@ -349,47 +349,42 @@ public override void Reset(DocumentChange change)
...
             _cts.Cancel();
 
-            var thread = _thread;
-            _thread = null;
+            var longRunningWork = _longRunningWork;
+            _longRunningWork = null;
 
-            if (Thread.CurrentThread != thread) // prevent a deadlock
-                thread.Join();
+            if (_longRunningWork != RavenThreadPool.LongRunningWork.Current) // prevent a deadlock
```
Comment (ayende) [inline src/Raven.Server/Documents/ETL/EtlProcess.cs:386]: You are comparing to the _field_ that you just set to null here.

File: src/Raven.Server/Utils/ThreadPool.cs [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+                pooled = new PooledThread(this);
+                var thread = new Thread(pooled.Run)
+                {
+                    Name = name,
+                    IsBackground = true,
+                };
+
+                thread.Start();
+            }
+            pooled.StartedAt = DateTime.Now;
```
Comment (ayende) [inline src/Raven.Server/Utils/ThreadPool.cs:80]: UtcNow

File: src/Raven.Server/Utils/ThreadPool.cs [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+                var runtimeThread = _runtimeThreadField.GetValue(t);
+                _threadFieldName = runtimeThread.GetType().GetField("m_Name", System.Reflection.BindingFlags.Instance | System.Reflection.BindingFlags.NonPublic);
+                Debug.Assert(_threadFieldName != null);
+                
+            }
+
+            public PooledThread(RavenThreadPool pool)
+            {                
+                _parent = pool;
+                _logger = LoggingSource.Instance.GetLogger<PooledThread>("PooledThread");
```
Comment (ayende) [inline src/Raven.Server/Utils/ThreadPool.cs:115]: Can probably use the parent logger

File: src/Raven.Server/Utils/ThreadPool.cs [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+                    Thread.CurrentThread.Name = _name;
+
+                    try
+                    {
+                        LongRunningWork.Current = _workIsDone;
+                        _action(_state);
+                    }
+                    finally
+                    {
+                        _workIsDone.Set();
```
Comment (ayende) [inline src/Raven.Server/Utils/ThreadPool.cs:194]: To be safe, let's reset the `Current` here.

File: src/Voron/Data/BTrees/Tree.cs [hunk 9]
```diff
@@ -63,7 +63,11 @@ private Tree(LowLevelTransaction llt, Transaction tx, long root, Slice name, boo
                 RootPageNumber = root
             };
         }
-
+        public Tree(bool iAmNotBoingToUseThisObjectUnlessIWantToCleanThreadStaticData)
```
Comment (ayende) [inline src/Voron/Data/BTrees/Tree.cs:66]: I don't like this

File: src/Raven.Server/Utils/ThreadPool.cs [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+                    ResetCurrentThreadName();
+                    Thread.CurrentThread.Priority = ThreadPriority.Normal;
+                    Thread.CurrentThread.Name = "Available Pool Thread";
+
+                    _waitForWork.WaitOne();
+                    _workIsDone.ManagedThreadId = Thread.CurrentThread.ManagedThreadId;
+                    if (_action == null)
+                        return; // should only happen when we shutdown
+
+                    ResetCurrentThreadName();
```
Comment (arekpalinski) [inline src/Raven.Server/Utils/ThreadPool.cs:184]: I wonder why do we need to reset it first before using `Thread.CurrentThread.Name = _name;`?

Comment (ayende) [reply src/Raven.Server/Utils/ThreadPool.cs:184]: It was set to `"Available Pool Thread"` on line 177, and it cannot be set to a value unless it is null

Comment (arekpalinski) [reply src/Raven.Server/Utils/ThreadPool.cs:184]: Good to know. Thanks

**Pull request merged**